source/azure_iot_hub_client.c

```diff
@@ -0,0 +1,582 @@
+                                                       AzureIoTHubClient_t * pxAzureIoTHubClient,
+                                                       void * pxPublishInfo )
+{
+    AzureIoTHubClientTwinResponse_t xTwinResponse;
+    AzureIoTMQTTPublishInfo_t * xMQTTPublishInfo = ( AzureIoTMQTTPublishInfo_t * ) pxPublishInfo;
+    az_iot_hub_client_twin_response xOutRequest;
+    az_span xTopicSpan = az_span_create( ( uint8_t * ) xMQTTPublishInfo->pTopicName, xMQTTPublishInfo->topicNameLength );
+    uint32_t xRequestID = 0;
+
+    /* Failed means no topic match */
+    if( az_result_failed( az_iot_hub_client_twin_parse_received_topic( &pxAzureIoTHubClient->_internal.xAzureIoTHubClientCore,
+                                                                       xTopicSpan, &xOutRequest ) ) )
+    {
+        return eAzureIoTHubClientTopicNoMatch;
+    }
+
+    /* Process incoming Publish. */
+    AZLogInfo( ( "Twin Incoming QoS : %d.", xMQTTPublishInfo->qos ) );
+
+    /* Verify the received publish is for the feature we have subscribed to. */
+    AZLogInfo( ( "Twin topic: %.*s. with payload : %.*s.",
+                 xMQTTPublishInfo->topicNameLength,
+                 xMQTTPublishInfo->pTopicName,
+                 xMQTTPublishInfo->payloadLength,
+                 xMQTTPublishInfo->pPayload ) );
+
+    if( pxContext->_internal.callbacks.xTwinCallback )
+    {
+        if( az_span_size( xOutRequest.xRequestID ) == 0 )
+        {
+            xTwinResponse.messageType = eAzureIoTHubTwinDesiredPropertyMessage;
+        }
+        else
+        {
+            if( az_result_failed( az_span_atou32( xOutRequest.xRequestID, &xRequestID ) ) )
+            {
+                return eAzureIoTHubClientFailed;
+            }
+
+            if( xRequestID & 0x01 )
+            {
+                xTwinResponse.messageType = eAzureIoTHubTwinReportedResponseMessage;
+            }
+            else
+            {
+                xTwinResponse.messageType = eAzureIoTHubTwinGetMessage;
+            }
+        }
+
+        xTwinResponse.pvMessagePayload = xMQTTPublishInfo->pPayload;
+        xTwinResponse.ulPayloadLength = ( uint32_t ) xMQTTPublishInfo->payloadLength;
+        xTwinResponse.messageStatus = xOutRequest.status;
+        xTwinResponse.pucVersion = az_span_ptr( xOutRequest.version );
+        xTwinResponse.usVersionLength = ( uint16_t ) az_span_size( xOutRequest.version );
+        xTwinResponse.ulRequestID = xRequestID;
+        pxContext->_internal.callbacks.xTwinCallback( &xTwinResponse,
+                                                      pxContext->_internal.pvCallbackContext );
+    }
+
+    return eAzureIoTHubClientSuccess;
+}
+/*-----------------------------------------------------------*/
+
+/**
+ *
+ * Time callback for MQTT initialization.
+ *
+ * */
+static uint32_t prvGetTimeMs( void )
+{
+    TickType_t xTickCount = 0;
+    uint32_t ulTimeMs = 0UL;
```

Comment:
>= 0UL [](start = 22, length = 5)

not required these initializations

Reply:
how about just 

static uint32_t prvGetTimeMs( void )
{
return xTaskGetTickCount() * azureiotMILLISECONDS_PER_TICK;
}


---
In reply to: [609119426](https://github.com/Azure/azure-iot-middleware-freertos/pull/60#discussion_r609119426) [](ancestors = 609119426)